### PR TITLE
Firmware sets

### PR DESCRIPTION
**Pull request by joelrebel**
This PR adds CRUD methods for firmware sets.

Firmware sets enable grouping of firmware objects and enforces a few patterns of firmware management
- firmware part of a set cannot be removed unless it's removed from the group.
- firmware sets cannot contain duplicate references
- firmware sets cannot reference multiple firmware versions of the same - `vendor`, `model`, `component` attributes.

Along with the points above, a firmware set acts as a release mechanism for firmwares available for installation.

In the next iteration of changes, firmware sets are to provide a labels field,
the labels field will enable firmware sets to be associated with one or more server, metadata attributes.

### CODE REVIEW

File: pkg/api/v1/router_firmware_set_test.go [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+		},
+		{
+			"404 returned for unknown firmware set UUID",
+			uuid.New(),
+			true,
+			"404",
+			"resource not found",
+		},
+		{
+			"update an existing firmware set - update name, referenced firmware",
```
Comment (diogomatsubara) [inline pkg/api/v1/router_firmware_set_test.go:368]: Looks like this test name is wrong.

File: pkg/api/v1/router_firmware_set_test.go [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+
+	testCases := []struct {
+		testName                    string
+		params                      *serverservice.ComponentFirmwareSetListParams
+		expectedFirmwareSetMetadata json.RawMessage
+		expectedFirmwareModel       string
+		expectedFirmwareCount       int
+		expectedTotalRecordCount    int
+		expectedPage                int
+		expectedErorr               bool
```
Comment (diogomatsubara) [inline pkg/api/v1/router_firmware_set_test.go:420]: Typo in the var name: s/expectedErorr/expectedError/

**Review by diogomatsubara — DISMISSED** [review]
Thanks for working on this. It looks good to me but I don't have permission to approve PRs in this project. I left a few minor comments.

**Comment by joelrebel** [comment]
Thanks for the feedback @diogomatsubara.

@nicolerenee can you take a look here?

**Comment by joelrebel** [comment]
FYI

In a following PR, I intend swap the `component_firmware_set` table `Metadata` column with Attributes instead.

https://github.com/metal-toolbox/hollow-serverservice/pull/149/commits/98e5362ac7786db238a65fe1b89549147efaf8b2#diff-448bfc1568bcc674243ab08f20fc824bd907d8511b240f254476d29013f523b9R7

**Comment by joelrebel** [comment]
for the record, the changes for firmware set attributes are here https://github.com/joelrebel/hollow-serverservice/tree/firmware-sets-attributes

**Comment by joelrebel** [comment]
Based on an internal discussion with @nicolerenee, firmware sets is to have its own Attributes table.
For now I'll close this PR and create a new one with the Attributes changes included.

**Pull request closed without merging**